### ISSUE TITLE
[false-positive] new-without-default lint should not report if there's a type parameter

### ISSUE THREAD

**Issue by torkleyy** [Manishearth/rust-clippy#1809]
I got a convenience function which looks like this:

```rust
pub fn new<T: Any>() -> Self {
    Self::from_type_id(TypeId::of::<T>())
}
```

Clippy shouldn't give me a warning here.

**Comment by oli-obk** [comment]
Wouldn't it be cleaner to relabel the method `from_type`?

**Comment by torkleyy** [comment]
While it's true that this name would be clearer, I still consider it a false positive.

**Comment by oli-obk** [comment]
Oh it totally is. That's why I marked it as a bug

**Comment by scott-linder** [comment]
I would be interested in taking this on. I added a test case to the lint and it does warn.

Should the lint ignore all `new<..>` with a generic type parameter?

**Comment by torkleyy** [comment]
@scott-linder I think so given that they're incompatible with `Default` in these cases.

**Comment by scott-linder** [comment]
@torkleyy for a simple enough test case it is possible to derive `Default` even with `new<..>()`, for example my test case is:

```rust
#[derive(Default)]
pub struct IgnoreGenericNew;
impl IgnoreGenericNew {
  pub fn new<T>() -> Self { IgnoreGenericNew } // the derived Default does not make sense here as the result depends on T
} 
```

However even for this case I don't think the lint should say anything, because although the derive is legal it is not necessarily sensible.

I already have a patch to just skip all `new()` methods which have any generic type parameters. I will open a PR with that and we can discuss it there.
